Reuse single VM for many sites

ethitter opened the pull request:
VMs consume a lot of space, and provisioning one from scratch is time consuming. In urgent situations particularly, the latter is unacceptable. Ideally, we can use a single VM for myriad sites, quickly switching between individual sites' git repos.

This PR introduces the basic support for quickly switching between discrete codebases and databases with little more than a call to `vagrant provision`. A git remote is specified and downloaded, and the VM reloaded with a database specific to the site. A bash script aids in these transitions, as an interim step until a comprehensive VIP Go CLI is available.

What works now?
- local checkout of the client's git repo, synced to the VM so you can follow your local development workflow
- each client has its own DB, allowing easy switching between discrete content, as well as reuse of databases when client slugs are consistent
- support for all standard vip-skeleton directories
- doesn't require `vagrant reload` or `vagrant destroy` to switch between sites
- sets some basic options to make the environment useful, like permalinks and a real timezone string

What doesn't it do yet?
- no content is imported
- as noted in #12, a site's theme and plugins aren't automatically activated
- reuse local checkouts; each init clones a fresh copy
- swap out uploads, so everything will end up in the same directories until `vagrant destroy`
- support custom domains for each client; the common http://vip-go.local/ domain is used

Why is the helper in Bash? For expediency at the moment, really. There's nothing particularly novel there, so no reason the fundamentals can't be ported to vip-cli at a later date. Some useful lessons were learned, which will apply regardless of which language certain Vagrant interactions are implemented in.

Why aren't custom domains allows? Due to the speed concerns noted at the outset. Changing hostnames requires reloading and re-provisioning (`vagrant reload --provision` or `vagrant destroy && vagrant up`), which is too slow. Sharing a single domain is a reasonable compromise for the speed we're afforded by only needing to call `vagrant provision` in the implementation herein.

How does one use this?

`./qs-init.sh --client UNIQUE_SLUG --git-repo GIT_REMOTE [--up]`

eg:
`./qs-init.sh --client vip-go-test.local --git-repo git@github.com:Automattic/vip-skeleton.git`

Include `--up` if you've never run `vagrant up` for this particular Quickstart instance.

ethitter committed:
Support multiple sites within a single VM

Use a dynamic value to create the VM configs so that multiple client sites can be served from a single VM instance.

TODO: fix zeroconf with each new provision
TODO: pull down client code and link to it as necessary, including vip-config.php

ethitter committed:
Use common domain for all sites, otherwise a reload is required

Speed is a primary concern as we'll need to set up the VM in emergency situations. Changing hostnames requires a `vagrant reload --provision`, which is too slow for those situations. Standardizing on `vip-go.local` speeds up the process of setting up the VM for a new client.

ethitter committed:
Update manifest to use new variables

Facter variables were changed in 2fef94f and this use was overlooked.

ethitter committed:
Correct host entry so that VM is actually accessible from the host machine.

As part of the changes in 7858c22, I broke the ability to connect to the VM from the host machine. Oops. Turns out, that's somewhat necessary. ;)

ethitter committed:
WordPress default options

Set some default options to make sites easier to differentiate when re-initing, and to cover some common settings that can result in weird inconsistencies between production and development.

ethitter committed:
Load client code into VM when re-initing.

With a properly-structured VIP Go repo checked out in `./go-client-repo/`, we can map the git repo structure to what WP expects.

TODO: vip-config/

ethitter committed:
Introduce simple bash script to accept required arguments and prepare the VM

ethitter committed:
Stop redirecting informational messages to `stderr`. Copy-pasta from the error lines.

ethitter committed:
Improve display of feedback from the bash wrapper

ethitter committed:
Hint init script so others have an idea how to use it.

ethitter committed:
Load client's `vip-config.php`, completing support for the skeleton repo's supported elements

ethitter committed:
Minor syntax cleanup

joshbetz commented on Vagrantfile:
```diff
@@ -20,7 +22,9 @@ Vagrant.configure(2) do |config|
     puppet.module_path = "puppet/modules"
     puppet.facter = {
-    	"quickstart_domain" => config.vm.hostname
+        "hostname"          => config.vm.hostname,
```
Hostname should be a built-in fact. https://docs.puppetlabs.com/facter/latest/core_facts.html#hostname

ethitter replied:
Good catch. I should've actually switched to the `fqdn` fact, otherwise we don't end up with a TLD. Updates forthcoming.

joshbetz commented:
This is awesome. I'm wondering if we should go with vip-go.local or go-vip.local, given that we have go-vip.co for staging sites.

ethitter commented:
> I'm wondering if we should go with vip-go.local or go-vip.local, given that we have go-vip.co for staging sites.

Definitely! The symmetry is great. I'll make that change along with the `fqdn` fact.

ethitter committed:
Hostname updates

Use `go-vip.local` to mirror our `go-vip.co` staging domain. Also, leverage built-in Puppet facts rather than defining things ourselves. Props @joshbetz.
